### ISSUE TITLE
Ban or enforce arguments + parameters in IIFEs

### ISSUE THREAD

**Issue by graingert** [eslint/eslint#8844]
**Please describe what the rule should do:**

Often people converting legacy `script` style JS into modules use IIFEs to alias variables. Write a rule to ban these.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style 
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something X
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**
With ban:
```js
// bad
((global) => {
    let foo = 3;
    ham();
    return 7;
})(window);

// good
(() => {
    let global = window;
    let foo = 3;
    ham();
    return 7;
})();

// bad
((ham, spam, undef) => {
    let foo = 3;
    ham();
    return 7;
})(eggs, chips);

// good
(() => {
    let ham = eggs;
    let spam = chips;
    let undef = void 0;
    let foo = 3;
    ham();
    return 7;
})();
```
with enforce:
```js
// bad
(() => {
    let global = window;
    let foo = 3;
    ham();
    return 7;
})();

// good
((global, foo) => {
    ham();
    return 7;
})(window, 3);

// bad
(() => {
    let ham = eggs;
    let spam = chips;
    let undef = void 0;
    let foo = 3;
    ham();
    return 7;
})();

// good
((ham, spam, foo, undef) => {
    ham();
    return 7;
})(eggs, chips, 3);

// bad
let bar = 4;
(() => {
    let ham = eggs;
    let spam = chips;
    let undef = void 0;
    let foo = 3;
    let zap = bar;
    ham(bar);
    return 7;
})();

// good
let bar = 4;
((ham, spam, foo, zap, undef) => {
    ham(bar);
    return 7;
})(eggs, chips, 3, bar);
```

**Why should this rule be included in ESLint (instead of a plugin)?**
Seems pretty handy, doesn't depend on exotic syntax.

**Comment by platinumazure** [comment]
Not sure about this. Lots of folks prefer the style of not closing over any variables (e.g., jQuery). I don't think ESLint core should endorse one style or another here.

I recommend writing a plugin with this rule.

**Comment by graingert** [comment]
@platinumazure ok updated with ban/enforce toggle.

**Comment by platinumazure** [comment]
This is pretty interesting now. Maybe if there were support for patterns (regular expressions) of allowed arguments, then users could customize for their use case?

Withdrawing my :-1:, would be interested to see what the team thinks.

**Comment by graingert** [comment]
>  Lots of folks prefer the style of not closing over any variables

To be clear I favour the 'ban' option of this rule.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
